### PR TITLE
feat: add validation for accept-withdrawal-request transactions

### PR DESCRIPTION
**Pull request by djordon**
## Description

Closes https://github.com/stacks-network/sbtc/issues/477.

## Changes

The changes here are very similar in spirit to the changes in https://github.com/stacks-network/sbtc/pull/545.

* Add a `SignerVotes` type for converting to a bitmap
* Implement validate for the `AcceptWithdrawalV1` type, which is used for `accept-withdrawal-request` contract calls.
* Minor clean-up of the integration tests for the `complete-deposit` validation integration tests.

## Testing Information

This PR adds integration tests. There is still some work to what is here before it is finalize, specifically https://github.com/stacks-network/sbtc/issues/552 and https://github.com/stacks-network/sbtc/issues/554.

## Checklist:

- [x] I have performed a self-review of my code

### CODE REVIEW

File: signer/src/stacks/contracts.rs [hunk 1]
```diff
@@ -518,20 +531,226 @@ impl AsContractCall for AcceptWithdrawalV1 {
...
+        }
+        // 4. That the sweep transaction has the UTXO indicated by the
+        //    outpoint.
+        //
+        // Okay great, we know that the sweep transaction exists on the
+        // canonical bitcoin blockchain, we just need to do a simple check
+        // of the transaction inputs.
+        sweep_tx
+            .output
+            .get(self.outpoint.vout as usize)
```
Comment (cylewitruk) [inline signer/src/stacks/contracts.rs:673]: [Question] Maybe a dumb Q, but is this check sufficient? I mean simply ensuring that the tx contains an output at the expected index? Does that output need to be checked somehow?

File: signer/src/stacks/contracts.rs [hunk 2]
```diff
@@ -518,20 +531,226 @@ impl AsContractCall for AcceptWithdrawalV1 {
...
     ///
-    /// 1. That the transaction with the associated request_id is stored as
-    ///    an event on the canonical Stacks blockchain.
-    /// 2. That the transaction associated with the outpoint has been
-    ///    confirmed on the canonical bitcoin blockchain.
-    /// 3. That the signer bitmap matches the signer decisions stored in
-    ///    this signer's database.
-    /// 4. That the `tx_fee` matches the amount spent to the bitcoin miner.
-    async fn validate<S>(&self, _db: &S, _ctx: &ReqContext) -> Result<(), Error>
+    /// 1. That the smart contract deployer matches the deployer in our
```
Comment (cylewitruk) [inline signer/src/stacks/contracts.rs:534]: [Suggestion] Pretty much the same doc comments I had on the other PR :)

File: signer/src/error.rs [hunk 3]
```diff
@@ -300,6 +301,11 @@ pub enum Error {
     #[error("unexpected public key from signature. key {0}; digest: {1}")]
     UnknownPublicKey(crate::keys::PublicKey, secp256k1::Message),
 
+    /// The error for when the request to sign a withdrawal-accept
+    /// transaction fails at the validation step.
+    #[error("{0}")]
+    WithdrawalAcceptValidation(#[source] Box<WithdrawalAcceptValidationError>),
```
Comment (cylewitruk) [inline signer/src/error.rs:307]: Does this need to be boxed? It's a struct local to this crate?

**Review by cylewitruk — COMMENTED** [review]
Haven't gotten all the way through yet but here's my comments so far :)

File: signer/src/stacks/contracts.rs [hunk 4]
```diff
@@ -518,20 +531,226 @@ impl AsContractCall for AcceptWithdrawalV1 {
...
+            .ok_or_else(|| WithdrawalErrorMsg::RequestMissing.into_error(ctx, self))?;
+
+        // 5. The `scriptPubKey` of the UTXO matches the one in the withdrawal
+        //    request.
+        if &tx_out.script_pubkey != request.recipient.deref() {
+            return Err(WithdrawalErrorMsg::RecipientMismatch.into_error(ctx, self));
+        }
+        // 6. The `amount` of the UTXO matches the one in the withdrawal
+        //    request.
+        if tx_out.value.to_sat() != request.amount {
```
[the remark this answers is not in the document]
Comment (djordon) [reply signer/src/stacks/contracts.rs:607]: It should always be true. There are two "amount" values in the withdrawal request, the `amount` that is spent to the intended recipient, and the `max-fee` to spend as a bitcoin transaction fee. Both amounts are locked as part of the withdrawal flow when the user initiates a withdrawal request.

File: signer/src/stacks/contracts.rs [hunk 5]
```diff
@@ -518,20 +531,226 @@ impl AsContractCall for AcceptWithdrawalV1 {
...
+        // From the above, we know that the sweep transaction is in the
+        // `sweep_block_hash`. Now we just need to check that this block is
+        // on the canonical bitcoin blockchain.
+        let block_ref = BitcoinBlockRef {
+            block_hash: self.sweep_block_hash,
+            block_height: self.sweep_block_height,
+        };
+
+        let in_canonical_bitcoin_blockchain = db
+            .in_canonical_bitcoin_blockchain(&ctx.chain_tip, &block_ref)
```
[the remark this answers is not in the document]
Comment (djordon) [reply signer/src/stacks/contracts.rs:660]: Oh yeah that is better.

File: signer/src/stacks/contracts.rs (hunk 1 above)
Comment (djordon) [reply signer/src/stacks/contracts.rs:673]: We check that it is being spent to the intended recipient and that it has the correct amount. Both should be checked, and if they are wrong then we did something wrong or someone is acting maliciously.

File: signer/src/stacks/contracts.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (cylewitruk) [reply signer/src/stacks/contracts.rs:607]: Ah okay, thx

File: signer/src/error.rs (hunk 3 above)
Comment (cylewitruk) [reply signer/src/error.rs:307]: Ah gotcha, so better to have it on the heap

File: signer/src/stacks/contracts.rs (hunk 2 above)
Comment (djordon) [reply signer/src/stacks/contracts.rs:534]: Done in https://github.com/stacks-network/sbtc/pull/555/commits/002fc518b6f1e828b0c2ab885f827cdc34baaa7d.

File: signer/src/bitcoin/utxo.rs [hunk 6]
```diff
@@ -348,32 +347,12 @@ impl DepositRequest {
     pub fn from_model(
         request: model::DepositRequest,
         signers_public_key: XOnlyPublicKey,
-        mut votes: Vec<SignerVote>,
+        votes: SignerVotes,
     ) -> Self {
-        let txid = request.txid.into();
-        let vout = request.output_index;
-
-        let mut signer_bitmap = BitArray::ZERO;
```
[the remark this answers is not in the document]
Comment (djordon) [reply signer/src/bitcoin/utxo.rs:356]: Yup, 👍🏽

File: signer/src/stacks/contracts.rs (hunk 1 above)
Comment (cylewitruk) [reply signer/src/stacks/contracts.rs:677]: Ahh right this tx is returned and passed into validate_utxo which does that. Thx, was reading top-down and got mixed up 😅

**Review by cylewitruk — COMMENTED** [review]
Still have a few tests left to go through but here's a few more comments (same stuff as on the other PR pretty much)

File: signer/tests/integration/withdrawal_accept.rs [hunk 7]
```diff
@@ -0,0 +1,992 @@
...
+    // our database.
+    let bitcoin_tx_ref = BitcoinTxRef {
+        txid: sweep_tx.txid.into(),
+        block_hash: sweep_tx.block_hash.into(),
+    };
+    db.write_transaction(&sweep_tx).await.unwrap();
+    db.write_bitcoin_transaction(&bitcoin_tx_ref).await.unwrap();
+
+    // Normal: get the signer bitmap for how they voted.
+    let bitmap = db
```
[the remark this answers is not in the document]
Comment (djordon) [reply signer/tests/integration/withdrawal_accept.rs]: Good call, https://github.com/stacks-network/sbtc/pull/555/commits/92306a461fa8517f6ae0754319f3502136f1ecb3.

File: signer/tests/integration/withdrawal_accept.rs [hunk 8]
```diff
@@ -0,0 +1,992 @@
...
+    let (aggregate_key, signer_set) = withdrawal_setup(&mut rng, &db, signatures_required).await;
+    // Get the chain tip.
+    let chain_tip = get_bitcoin_canonical_chain_tip_block(&db)
+        .await
+        .unwrap()
+        .unwrap();
+    // Normal: Get an existing withdrawal request on the canonical bitcoin
+    // blockchain.
+    let req = db
+        .get_pending_accepted_withdrawal_requests(&chain_tip.block_hash, 20, signatures_required)
```
[the remark this answers is not in the document]
Comment (djordon) [reply signer/tests/integration/withdrawal_accept.rs]: Another good call, https://github.com/stacks-network/sbtc/pull/555/commits/92306a461fa8517f6ae0754319f3502136f1ecb3.

File: signer/tests/integration/withdrawal_accept.rs [hunk 9]
```diff
@@ -0,0 +1,992 @@
...
+    // transactions in each bitcoin block.
+    let db_num = DATABASE_NUM.fetch_add(1, Ordering::SeqCst);
+    let db = testing::storage::new_test_database(db_num, true).await;
+    let mut rng = rand::rngs::StdRng::seed_from_u64(51);
+    let signatures_required = 3;
+
+    let (aggregate_key, signer_set) = withdrawal_setup(&mut rng, &db, signatures_required).await;
+    // Get the chain tip.
+    let chain_tip = get_bitcoin_canonical_chain_tip_block(&db)
+        .await
```
[the remark this answers is not in the document]
Comment (djordon) [reply signer/tests/integration/withdrawal_accept.rs]: Oh right. Done, https://github.com/stacks-network/sbtc/pull/555/commits/92306a461fa8517f6ae0754319f3502136f1ecb3.

**Review by cylewitruk — APPROVED** [review]
lgtm :partying_face:

**Pull request merged**